### DISCUSSION TITLE
Authentication of Api

### DISCUSSION THREAD

**Discussion by florian30491** [acacode/swagger-typescript-api#1133] [Q&A]
Hi,

I have a problem to use the typescript api generator. 

In the examples of the api generator there is a possibility to set username and password for the api, see image below.

![image](https://user-images.githubusercontent.com/58624223/155694105-f0e816ce-c4ba-461e-9177-90d9a182991c.png)

Unfortunatly my api does not show how these options with the authentication login, see image below.

![image](https://user-images.githubusercontent.com/58624223/155694233-32962fe9-baa1-4598-b081-10c68c3bba96.png)

I can just set some securityData options, but for me it is not clear how to pass the security worker options. Especially with username and password.

Maybe someone can help me here ? 

Thanks in advance.

BR,
Florian

**Comment by mackehansson** [comment]
I may be wrong here but the method login in the example belongs on the auth? In your case it should be on the v3 property? api in example is v3 in your case.

**Comment by js2me** [comment]
Hi @florian30491 !   
Example from the first image it is single http file with login namespace routes   
Login and authentication basically depends on your swagger schema.   
swagger-typescript-api provides flag secure which calls specific method `securityWorker`, but you can modify this too

**Comment by js2me** [comment]
also @mackehansson @florian30491 better to remove v3 from your api and use `--module-name-index` which means   
If your request paths looks like this   
`/api/v3/fruits`   
`/api/v3/apples`  
`/api/v3/cookies`   

By default swagger-typescript-api will generate single namespace for routes with name `api` (`api - 0, v3 - 1, cookies/apples/etc - 2`)    

0 1 2 that is values for flag `--module-name-index`

**Comment by js2me** [comment]
`setSecurityData` is needed to set some specific auth data for your security requests. example   

const result = api.login.auth() // not secure, return token for all other secure requests   

api.setSecurityData(result.token)   

and each http client provides the `securityWorker` which is needed to process your "auth data"   

![image](https://user-images.githubusercontent.com/16340911/202910173-0959ae7b-baa8-4b38-bcc0-d891277e56a7.png)


```
const api = new Api({
   securityWorker: (token) => ({ headers: { 'Auth': token } })
})
```
